packages/encoder/lib/encoders.ts

```diff
@@ -0,0 +1,1302 @@
+   * type-value input as discussed in the documentation for [[wrap]], or you
+   * can skip overload resolution and explicitly select an overload by other
+   * means.  For enums you may also specify the enum type as documented in
+   * [[wrap]].
+   *
+   * @param methods An array of [[Method]] objects we are
+   *   attempting to decide between.
+   * @param inputs An array of the inputs to the transaction.  May include a
+   *   transaction options argument on the end if the `allowOptions` flag is
+   *   set.
+   * @param options Contains options to control the operation of this method.
+   * @return The interpretation of the input and the resolved method, as a
+   *   [[Resolution]] object.
+   */
+  public async resolveAndWrap(
+    methods: Codec.Wrap.Method[],
+    inputs: unknown[],
+    options: Types.ResolveOptions = {}
+  ): Promise<Codec.Wrap.Resolution> {
+    return await this.driveGenerator(
+      Codec.Wrap.resolveAndWrap(methods, inputs, {
+        userDefinedTypes: this.userDefinedTypes,
+        allowOptions: Boolean(options.allowOptions)
+      })
+    );
+  }
+
+  private async driveGenerator<T>(
+    generator: Generator<Codec.WrapRequest, T, Codec.WrapResponse>
+  ): Promise<T> {
+    let response: Codec.WrapResponse;
+    let next = generator.next();
+    while (!next.done) {
+      const request = next.value;
+      debug("request: %O", request);
+      // @ts-ignore: HACK HACK to make typedoc work
+      // (the TS is fine with strict null checks on,
+      // but typedoc has it turned off, so... :-/ )
+      // please remove this ts-ignore once you turn on
+      // strict null checks in typedoc
+      response = await this.respond(request);
+      debug("response: %O", response);
+      next = generator.next(response);
+    }
+    debug("returning: %O", next.value);
+    return next.value;
+  }
+
+  private async respond(
+    request: Codec.WrapRequest
+  ): Promise<Codec.WrapResponse> {
+    switch (request.kind) {
+      case "integer":
+        return this.recognizeInteger(request.input);
+      case "decimal":
+        return this.recognizeDecimal(request.input);
+      case "address":
+        return await this.recognizeAddress(request.name);
+    }
+  }
+
+  /**
+   * **This method is asynchronous.**
+   *
+   * This method is similar to [[wrapForTransaction]], except instead of
+   * returning a [[Resolution]] object, it returns
+   * the resulting transaction options, including the encoded `data`.
+   * Note any options not specified in a transaction options parameter
+   * will be simply omitted; it is up to the caller to set these as
+   * appropriate afterwards.
+   *
+   * If the transaction options parameter has a `data` option, this option will
+   * be recognized but ignored.  Similarly, when encoding a contract creation,
+   * the `to` option will be similarly ignored.
+   *
+   * See [[wrapForTransaction]] for documentation of the inputs.
+   */
+  public async encodeTransaction(
```

Comment:
As discussed, I think this `encodeTransaction` method should be considered the primary interface - likely we want to `s/resolveAndEncode/encodeTransaction` (renaming this method to something else).

As part of this, we ought to accept function name strings in addition to `Method` objects directly.

Reply:
OK, I added `forAddress`.  I haven't tested it though.  Also, it's largely copypasted from the decoder.  It doesn't accept a block parameter though, it just uses `"latest"`.

I also haven't done the renaming yet.

Reply:
OK, did that renaming!  @gnidan, still waiting for answers about everything else.

Reply:
> I noticed that the Decoder, if you try to spawn a ContractDecoder by name, but it has no bytecode, it'll do a lookup by name. D: Yikes! And of course the Encoder now copies that behavior... are we OK with that for now, or does that need changing?

I say that this probably needs changing. Lookup by name is blurgh, too risky.

> Note: This opens the question of whether we want a special error for when the user puts in a string that doesn't actually match any function name.

Seems reasonable!

> OK, I added `forAddress`. I haven't tested it though. Also, it's largely copypasted from the decoder. It doesn't accept a block parameter though, it just uses "latest".

Any way to make it accept a block number?

(Any other points I missed here?)

Reply:
> I say that this probably needs changing. Lookup by name is blurgh, too risky.

Yeah, still need to figure out what to do about this.  We'll talk more about this I guess.

> Seems reasonable!

Fortunately, I've already done this! :)

> Any way to make it accept a block number?

IIRC, the reason I didn't do so is because the adapter function for getting code is typed as taking a regularized block specifier, and I didn't want to drag in the whole regularization mechanism here.  However, I *already* just hackily changed that type, so... yeah, why not, I guess I can.  Of course it'll have to be a block *number* (or one of the specific allowed strings), rather than a block hash, but I guess that's already how decoder works.

Reply:
OK, I've changed it to allow an optional block number.

Reply:
Re: project encoder concerns, let's go with a stopgap solution of marking this method internal-only.

Reply:
OK, I hid those project encoder methods, and adjusted other documentation appropriately.

While I was at it, I realized that not all of the things that ought to allow taking a function name did, so I made that possible, and factored the code that does it.

I'm realizing now though that the documentation is written in a way that doesn't really focus things around `encodeTransaction` like they probably should, but I'll deal with that later?

Reply:
Also I still need to deal with the name lookup problem.

Reply:
OK, after discussing things with @gnidan, I changed things so that it will still look up by name if there's a unique name match (note this is only if there's no bytecode match of course!), but otherwise, rather than failing, will return a minimal default result.  However, in all cases, we then trust the passed-in ABI rather than the found ABI.

Note I didn't add any tests of the default result case.